numpyro/distributions/transforms.py

```diff
@@ -0,0 +1,157 @@
+# The implementation follows the design in PyTorch: torch.distributions.transforms.py
+
+import jax.numpy as np
+from jax.scipy.special import expit, logit
+
+
+class Transform(object):
+    domain = constraints.real
+    codomain = constraints.real
+    event_dim = 0
+
+    def __call__(self, x):
+        return NotImplementedError
+
+    def inv(self, y):
+        raise NotImplementedError
+
+    def log_abs_det_jacobian(self, x, y):
+        raise NotImplementedError
+
+
+# TODO: currently, just consider dim=0 or dim=1
+def _sum_rightmost(x, dim):
+    return np.sum(x, axis=-1) if dim == 1 else x
+
+
+class ComposeTransform(Transform):
+    def __init__(self, parts):
+        self.parts = parts
+
+    @property
+    def domain(self):
+        return self.parts[0].domain
+
+    @property
+    def codomain(self):
+        return self.parts[-1].codomain
+
+    @property
+    def event_dim(self):
+        return max(p.event_dim for p in self.parts)
+
+    def __call__(self, x):
+        for part in self.parts:
+            x = part(x)
+        return x
+
+    def inv(self, y):
+        for part in self.parts[::-1]:
+            y = part.inv(y)
+        return y
+
+    def log_abs_det_jacobian(self, x, y):
+        result = 0
+        for part in self.parts[:-1]:
+            y_tmp = part(x)
+            result = result + _sum_rightmost(part.log_abs_det_jacobian(x, y_tmp),
+                                             self.event_dim - part.event_dim)
+            x = y_tmp
+        result = result + _sum_rightmost(part.log_abs_det_jacobian(x, y),
+                                         self.event_dim - self.parts[:-1].event_dim)
+        return result
+
+
+class AffineTransform(Transform):
+    # TODO: currently, just support scale > 0
+    def __init__(self, loc, scale, domain=constraints.real):
+        self.loc = loc
+        self.scale = scale
+        self.domain = domain
+
+    @property
+    def codomain(self):
+        if self.domain is constraints.real:
+            return constraints.real
+        elif isinstance(self.domain, constraints.greater_than):
+            return constraints.greater_than(self.loc + self.scale * self.domain.lower_bound)
+        elif isinstance(self.domain, constraints.interval):
+            return constraints.interval(self.loc + self.scale * self.domain.lower_bound,
+                                        self.loc + self.scale * self.domain.upper_bound)
```

Comment:
this way, we can support AffineTransform for more classes of distributions. Currently, the one in PyTorch just supports real domain.

Reply:
Nice, we should also add this to PyTorch!

Reply:
PyTorch licence includes [many copyrights](https://github.com/pytorch/pytorch/blob/master/LICENSE). Which one would you suggest?